doc/sla.md

```diff
@@ -0,0 +1,719 @@
+- set of fingerprints
+- the dependency model
+- the consumer signature
+
+The SLA contract will emit events for the service id. This will indicate that the SLA setup status is true.
+
+
+**3. Execute Phase**
+
+Once the SLA is publicly available on-chain, the actors will trigger the associated tranasctions. For instance, the consumer will 
+lock the payment on-chain `fulfills lockPayment` condition. An event will be emitted by the condition indicating a change happened on-chain, the 
+marketplace will respond by granting access to the user/actor-n and so on. This seems like a chain reaction between 
+parties. 
+
+**4. Dispute Phase (Optional)**
+
+Dispute in merkelized SLA is super easy because, it could be resolved by `log(n)` search in the merkelized SLA. The dispute itself could be defined 
+as a part of the SLA conditions as well where this condition points to the controller contract that is responsible for new dispute such as running verification game through `trubit network`.
+
+
+**5. Fulfilling Phase**
+
+Fulfilling service level agreement is managed by fulfilling all the conditions in the on-chain instance of the service agreement.
+
+## Marketplace Reference Architecture
+
+The marketplace is one of the Ocean protocol actors, the following figure proposes the reference components where a marketplace
+might need in order to interact with the ocean network.
+
+![reference marketplace](img/SLA_marketplace.png)
+
+The frontend is use-case specific and it is based on the published service. The gray components 
+are provided by ocean protocol, the blue components are marketplace use-case specific.
+
+## Appendix
+
+### Function Fingerprint
+
+In solidity, the function fingerprint is defined by the the function signature. The signature is 
+based on the `function selector` and `argument encoding`. The following example shows how to get 
+the function fingerprint:
+
+```javascript
+pragma solidity ^0.4.24;
+
+contract Foo {
+  function baz(uint32 x, bool y) public pure returns (bool r) { r = x > 32 || y; }
+}
+
+```
+
+We can drive the fingerprint of `baz(uint32, bool)` by calculating the `Keccak-256` of the function then
+get the first 4 bytes of  hash `0xcdcd77c0992ec5bbfc459984220f8c45084cc24d9b6efed1fae540db8de801d2`of the ASCII form of the signature: `0xcdcd77c0`. This what we need to add as a fingerprint for the function. For more information, check out the 
+[Solidity - ABI Function Selectors](https://solidity.readthedocs.io/en/develop/abi-spec.html#abi-function-selector). The function selector is one of special members in `public` and `external` functions. This selector returns the ABI function selector 
+as follows:
+
+```javascript
+pragma solidity ^0.4.25;
+contract Selector {
+  function f() public view returns (bytes4) {
+    return this.f.selector;
+  }
+}
+```
+
+### Controller-Storage Pattern
+
+The controller-storage pattern (also know as Interface, Controller, and Storage or ICS Pattern) is used as a design pattern
```

Comment:
Is this a universal design pattern for solidity contracts? Could you point out an external link where it is described in detail?

Reply:
Could you point to another implementation that documents this design?

Reply:
I will